light-client-verifier/src/operations/voting_power.rs

```diff
@@ -120,55 +164,135 @@ impl<V> Default for ProvidedVotingPowerCalculator<V> {
+    /// `None`.  Otherwise, if the vote is missing a signature returns
+    /// `Some(Err)`.  Otherwise, returns a `SignedVote` corresponding to given
+    /// `CommitSig`.
+    pub fn new(
+        commit_sig: &CommitSig,
+        validator_index: ValidatorIndex,
+        commit: &Commit,
+        chain_id: &chain::Id,
+    ) -> Option<Result<Self, VerificationError>> {
+        let (validator_address, timestamp, signature) = match commit_sig {
+            CommitSig::BlockIdFlagAbsent { .. } => return None,
+            CommitSig::BlockIdFlagCommit {
+                validator_address,
+                timestamp,
+                signature,
+            } => (*validator_address, *timestamp, signature),
+            CommitSig::BlockIdFlagNil { .. } => return None,
+        };
+
+        let vote = Vote {
+            vote_type: tendermint::vote::Type::Precommit,
+            height: commit.height,
+            round: commit.round,
+            block_id: Some(commit.block_id),
+            timestamp: Some(timestamp),
+            validator_address,
+            validator_index,
+            signature: signature.clone(),
+            extension: Default::default(),
+            extension_signature: None,
+        };
+        Some(
+            SignedVote::from_vote(vote, chain_id.clone())
+                .ok_or_else(VerificationError::missing_signature)
+                .map(|signed_vote| Self {
+                    signed_vote,
+                    verified: false,
+                }),
+        )
+    }
+
+    /// Returns address of the validator making the vote.
+    pub fn validator_id(&self) -> account::Id {
+        self.signed_vote.validator_id()
+    }
 }
 
-impl<'a> ValidatorMap<'a> {
-    /// Constructs a new map from given list of validators.
-    ///
-    /// Sorts the validators by address which makes the operation’s time
-    /// complexity `O(N lng N)`.
-    ///
-    /// Produces unspecified result if two objects with the same address are
-    /// found.  Unspecified in that it’s not guaranteed which entry will be
-    /// subsequently returned by [`Self::find_mut`] however it will always be
-    /// consistently the same entry.
-    pub fn new(validators: &'a [validator::Info]) -> Self {
-        let mut validators = validators.iter().map(|v| (v, false)).collect::<Vec<_>>();
-        validators.sort_unstable_by_key(|item| &item.0.address);
-        Self { validators }
+/// Collection of non-absent commit votes.
+struct NonAbsentCommitVotes {
+    /// Votes sorted by validator address.
+    votes: Vec<NonAbsentCommitVote>,
+}
+
+impl NonAbsentCommitVotes {
+    pub fn new(signed_header: &SignedHeader) -> Result<Self, VerificationError> {
+        let mut votes = signed_header
+            .commit
+            .signatures
+            .iter()
+            .enumerate()
+            .flat_map(|(idx, signature)| {
+                let idx = ValidatorIndex::try_from(idx).unwrap();
```

Comment:
Maybe add a comment explaining why this is infallible or use an infallible constructor?